File: apps/api/src/main/java/vnpt/vsp/module/tournament/entity/TournamentPolicyChange.java
package vnpt.vsp.module.tournament.entity;

import jakarta.persistence.*;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.type.SqlTypes;
import java.time.Instant;
import java.util.Map;
import java.util.UUID;

/**
 * Audit record for TournamentPolicy changes.
 *
 * Per PRD §8.12: "Enabled feature set is auditable."
 *
 * Every create, update, and lock operation emits a record with
 * before/after JSON snapshots for compliance review.
 */
@Entity
@Table(name = "tournament_policy_changes")
public class TournamentPolicyChange {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @Column(name = "policy_id", nullable = false)
    private UUID policyId;

    @Column(name = "changed_by", nullable = false)
    private Long changedBy;

    @Column(name = "changed_at", nullable = false)
    private Instant changedAt;

    @JdbcTypeCode(SqlTypes.JSON)
    @Column(name = "before_json", columnDefinition = "jsonb")
    private Map<String, Object> beforeJson;

    @JdbcTypeCode(SqlTypes.JSON)
    @Column(name = "after_json", nullable = false, columnDefinition = "jsonb")
    private Map<String, Object> afterJson;

    @Column(name = "reason", length = 500)
    private String reason;

    @PrePersist
    protected void onCreate() {
        if (changedAt == null) {
            changedAt = Instant.now();
        }
    }

    // ─── Getters / Setters ──────────────────────────────────────────────────

    public UUID getId() { return id; }
    public void setId(UUID id) { this.id = id; }

    public UUID getPolicyId() { return policyId; }
    public void setPolicyId(UUID policyId) { this.policyId = policyId; }

    public Long getChangedBy() { return changedBy; }
    public void setChangedBy(Long changedBy) { this.changedBy = changedBy; }

    public Instant getChangedAt() { return changedAt; }
    public void setChangedAt(Instant changedAt) { this.changedAt = changedAt; }

    public Map<String, Object> getBeforeJson() { return beforeJson; }
    public void setBeforeJson(Map<String, Object> beforeJson) { this.beforeJson = beforeJson; }

    public Map<String, Object> getAfterJson() { return afterJson; }
    public void setAfterJson(Map<String, Object> afterJson) { this.afterJson = afterJson; }

    public String getReason() { return reason; }
    public void setReason(String reason) { this.reason = reason; }
}

File: apps/mobile/lib/domain/models/tournament_policy_change.dart
// TournamentPolicyChange — Audit record for TournamentPolicy changes
// VSP Mobile App
//
// Tracks every policy creation and modification with before/after snapshots
// for compliance and auditability.
//
// Per PRD §8.12: "Enabled feature set is auditable."
//
// Story 7.4 — Slice A: TournamentPolicy Domain Model

import 'package:equatable/equatable.dart';

/// Audit record capturing a single change to a [TournamentPolicy].
///
/// Emitted on every create, update, and lock operation so that
/// tournament directors can review the full history of a policy.
class TournamentPolicyChange extends Equatable {
  /// Unique identifier for this audit record.
  final String id;

  /// ID of the policy that was changed.
  final String policyId;

  /// ID of the actor who made the change (user ID or 'system').
  final String changedBy;

  /// Timestamp when the change was made.
  final DateTime changedAt;

  /// JSON snapshot of the policy before this change.
  ///
  /// Null for policy creation events.
  final Map<String, dynamic>? beforeJson;

  /// JSON snapshot of the policy after this change.
  final Map<String, dynamic> afterJson;

  /// Optional reason for the change (e.g., "Round started", "Director update").
  final String? reason;

  const TournamentPolicyChange({
    required this.id,
    required this.policyId,
    required this.changedBy,
    required this.changedAt,
    this.beforeJson,
    required this.afterJson,
    this.reason,
  });

  /// Returns true if this is a creation event (no before state).
  bool get isCreation => beforeJson == null;

  /// Returns true if this is a lock event.
  bool get isLockEvent =>
      beforeJson != null &&
      afterJson['isLocked'] == true &&
      (beforeJson!['isLocked'] as bool? ?? false) == false;

  // ─── JSON ─────────────────────────────────────────────────────────────────

  factory TournamentPolicyChange.fromJson(Map<String, dynamic> json) {
    return TournamentPolicyChange(
      id: json['id'] as String,
      policyId: json['policyId'] as String,
      changedBy: json['changedBy'] as String,
      changedAt: DateTime.parse(json['changedAt'] as String),
      beforeJson: json['beforeJson'] as Map<String, dynamic>?,
      afterJson: json['afterJson'] as Map<String, dynamic>,
      reason: json['reason'] as String?,
    );
  }

  Map<String, dynamic> toJson() => {
    'id': id,
    'policyId': policyId,
    'changedBy': changedBy,
    'changedAt': changedAt.toIso8601String(),
    if (beforeJson != null) 'beforeJson': beforeJson,
    'afterJson': afterJson,
    if (reason != null) 'reason': reason,
  };

  @override
  List<Object?> get props => [
    id,
    policyId,
    changedBy,
    changedAt,
    beforeJson,
    afterJson,
    reason,
  ];
}
